Add incorrect workflow config notice to deployment logs

gcsecsey opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Fixes [DOTDEV-166 Add incorrect workflow config notice to deployment logs](https://linear.app/a8c/issue/DOTDEV-166/add-incorrect-workflow-config-notice-to-deployment-logs)

## Proposed Changes

This PR adds a notice when a deployment run fails due to an invalid workflow configuration. The notice appears when the logs contain a "Workflow validation failed" error message and provides a direct link to the deployment management page where users can fix the workflow file.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

For manually triggered deployments, this notice is already being shown when the validation fails. We added the same validation to automatic deployments in 184674-ghe-Automattic/wpcom, so I propose adding the same notice to this error for a consistent UX.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Pre-requisite: failing workflow

- Create a repository and connect it to one of your sites in an advanced configuration
- Push a change to your repository that breaks the workflow configuration, eg. update the name of the artifact to something else than wpcom

### Testing the changes

- Checkout this branch or use the live link
- Go to https://wordpress.com/github-deployments/:siteId and open the logs for the failed run
- Check that there's an error message about the failed validation
- Check that the notice is shown on the top right
- Check that the link redirects to the workflow configuration page

![CleanShot 2025-06-16 at 12 16 38@2x](https://github.com/user-attachments/assets/6f56d530-a69e-4b3a-93a6-8a90af8fc7ac)

gcsecsey committed:
add incorrect workflow config notice to error logs

gcsecsey committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into dotdev-166-add-notice-about-incorrect-workflow-config-to-deployment

ivan-ottinger reviewed (commented):
Thank you for the changes, Gergely!

The error noice is displayed as expected and the `Take action` link works correctly as well. 👍🏼 

One difference I see compared to the experience when the manually-triggered deploy fails is that the error notice gets triggered only when user navigates to the Deploy runs section and then opens up the logs section:

![Markup on 2025-06-16 at 14:48:39](https://github.com/user-attachments/assets/e2b1800a-f589-4421-b496-695088077af9)

This means that if the user is at `https://wordpress.com/github-deployments/{site-slug}`, they won't see the notice.

epeicher reviewed (approved):
Thanks for improving the notification! I have tested it, and it works as described but I was originally expecting the notification to be displayed automatically, related to what @ivan-ottinger describes [in his review](https://github.com/Automattic/wp-calypso/pull/104246#pullrequestreview-2932045178).

Also, everytime the user expands the error, the notification is displayed, which could be a bit annoying 🤓. What do you think @gcsecsey and @ivan-ottinger ? On one hand, having the `Take action` link that can be handy, on the other hand, having the notification always being displayed can be annoying 🤔

gcsecsey committed:
move notice to `/github-deployments/{site-slug}` page

gcsecsey commented:
> This means that if the user is at `https://wordpress.com/github-deployments/{site-slug}`, they won't see the notice.

Thanks for mentioning this @ivan-ottinger! My original assumption was that I won't be able to get the error message to detect if I need to show the notice on the `/github-deployments/{site-slug}` page. But when investigating this further, I found that the failure code is returned there, which can also be used for this purpose once 185605-ghe-Automattic/wpcom is landed. I moved the code triggering the notice to the `/github-deployments/{site-slug}` page now.

> Also, everytime the user expands the error, the notification is displayed, which could be a bit annoying 🤓. What do you think @gcsecsey and @ivan-ottinger ? On one hand, having the `Take action` link that can be handy, on the other hand, having the notification always being displayed can be annoying 🤔

Thanks @epeicher for this feedback! Upon testing this a bit further, I agree, it gets annoying that the notice is displayed each time the logs are toggled. For now I removed showing the notice from the logs page because of this, and only show it on the main deployments page. WDYT?

ivan-ottinger reviewed (approved):
Nice! Thank you for the updates, Gergely!

The changes work well:

![Markup on 2025-06-16 at 18:07:42](https://github.com/user-attachments/assets/e6b3a3f5-9630-4769-9764-1a5247fad2f1)

I have noticed only one edge case: When the error (and notice is triggered) already, we dismiss it and then do another commit with broken workflow file, the notice does not get triggered anymore.

I believe that might be because in that case `run?.failure_code` is still the same as before and hence the related `useEffect` does not run again.

---

Another thought that came to my mind is that maybe we could consider extracting out the contents of the `errorNotice` - since it is being used in a similar way for the other failed deploy case above (from line `53`). Just to reduce duplication.

epeicher commented:
>  I removed showing the notice from the logs page because of this, and only show it on the main deployments page. WDYT?

Thanks for iterating on this @gcsecsey, really appreciate the update! 🙏 

I think this might move the notification overload from one screen to another. Do you think there's a way to show it only once (or discard it after the first time) so users don’t see it every time they visit the page?

Sorry for being a bit picky here, just trying to make sure devs don’t get overwhelmed with repeated errors 🤓 and happy to get other views here!

wojtekn commented:
I triggered E2E tests.

gcsecsey committed:
re-trigger error notice for subsequent failures

gcsecsey committed:
extract error notice to a separate function

gcsecsey committed:
prevent showing notice on refetch

gcsecsey commented:
> Nice! Thank you for the updates, Gergely!
> 
> The changes work well:
> 
> ![Markup on 2025-06-16 at 18:07:42](https://private-user-images.githubusercontent.com/25105483/455619010-e6b3a3f5-9630-4769-9764-1a5247fad2f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAyNDE0ODUsIm5iZiI6MTc1MDI0MTE4NSwicGF0aCI6Ii8yNTEwNTQ4My80NTU2MTkwMTAtZTZiM2EzZjUtOTYzMC00NzY5LTk3NjQtMWE1MjQ3ZmFkMmYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjE4VDEwMDYyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2OWNhZDg4MmZkMzM0NWFlNjMzNjBmMGNmOThlYzIyMDNjOWRkZWRjZjI0MmY3NjE3NGIyZDI0MDUzZjkxMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.juoHVCdGKlqgahSJ7pgcwkE0rthJ-amxnc2uPdz45IQ)
> 
> I have noticed only one edge case: When the error (and notice is triggered) already, we dismiss it and then do another commit with broken workflow file, the notice does not get triggered anymore.
> 
> I believe that might be because in that case `run?.failure_code` is still the same as before and hence the related `useEffect` does not run again.
> 
> Another thought that came to my mind is that maybe we could consider extracting out the contents of the `errorNotice` - since it is being used in a similar way for the other failed deploy case above (from line `53`). Just to reduce duplication.

Thanks for these

ivan-ottinger reviewed (approved):
Thank you for the updates, Gergely.

The changes work as expected and the notice is displayed as it should be. Run IDs are stored in the local storage correctly:

![Markup on 2025-06-18 at 13:25:40](https://github.com/user-attachments/assets/c1945c85-e844-40f2-8237-19282f861f3f)

I did not observe any regressions.

epeicher reviewed (commented):
Thanks for iterating on this @gcsecsey! I have tested latest changes and they are great, the error is displayed without intervention of the user and when closing it. It is not displayed anymore. LGTM!

|Error happening | Closing the notifcation, navigating away and coming back to Deployments |
|-------|------|
| ![CleanShot 2025-06-18 at 14 46 21@2x](https://github.com/user-attachments/assets/6ae3d65e-a1ed-43dc-8717-d3c1738843b6)| ![CleanShot 2025-06-18 at 14 46 29@2x](https://github.com/user-attachments/assets/e413313e-3760-46b9-876c-a7c059041682)|
